[historical-view] Adds new timecontroller component

Kongkille opened the pull request:
TBD

Kongkille commented on web/src/hooks/redux.js:
```diff
@@ -16,10 +16,23 @@ export function useCurrentZoneHistory() {
   return useMemo(() => histories[zoneId] || null, [histories, zoneId]);
 }
 
+export function useCurrentDatetimes() {
+  // TODO: should use V5 state here and v5 should tell the state datetimes
+  const histories = useSelector((state) => state.data.histories);
+  if (histories && Object.keys(histories).length) {
+    return histories[Object.keys(histories)[0]].map((h) => moment(h.stateDatetime).toDate());
+  } else {
+    return [];
+  }
+}
```
This function assumes all histories have the same datetimes. As the comment suggests, V5/state should keep track of the relevant datetimes instead of the individual zones

I don't believe we have any cases where it makes sense for some zones to have a different set of datetimes than another.

Kongkille commented on web/src/reducers/index.js:
```diff
@@ -49,6 +49,7 @@ const initialApplicationState = {
   solarColorbarValue: null,
   webGLSupported: true,
   windColorbarValue: null,
+  selectedTimeAggregate: 'day',
```
Arguably this could be called 'selectedTimeRange' since the aggregate is 'hourly'. We could wait for another PR as we will touch this again soon.

madsnedergaard replied:
Yeah, I think this is fine for now and then we'll think more about it later on

Kongkille commented on web/src/components/graph/timeaxis.js:
```diff
@@ -20,15 +36,24 @@ const getTicksValuesFromTimeScale = (scale, count) => {
   const endTime = scale.domain()[1].valueOf();
 
-  const precision = moment.duration(15, 'minutes').valueOf();
+  const precision = moment.duration(60, 'minutes').valueOf();
```
This change should not affect anything, as we only have hourly values now

madsnedergaard replied:
FYI: This has changed in #4070

Kongkille commented:
> > @madsnedergaard marking this ready for review. I'll clear out the conflicts, I believe they are mostly related to lodash and moment
> 
> Perhaps it's time to get rid of both of them then! #4070 and #3985 will do that for you! 😉

Haha yes, you are right. I actually thought those were in! I'll get them merged next :)

VIKTORVAV99 commented on web/public/locales/en.json:
```diff
@@ -49,6 +49,13 @@
     },
+    "time-controller": {
+        "title": "Display data from the last",
+        "day": "Day",
+        "month": "Month",
+        "year": "Year",
+        "5year": "5 years"
```
Consider using `Intl.DisplayNames()` and Intl.NumberFormat() for these instead then we don't need to include them in the translation files.

Code needed:
```js
new Intl.DisplayNames(document.documentElement.lang, {type: 'dateTimeField'}).of('day')
```
and:
```js
new Intl.NumberFormat(document.documentElement.lang,  { style: 'unit', unit: 'year', unitDisplay: 'long' }).format(5);
```
[MDN | Intl.DisplayNames() constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DisplayNames/DisplayNames#using_type_datetimefield)
[MDN | Intl.NumberFormat() constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat)

madsnedergaard replied:
I agree, that would be a nice approach :)

VIKTORVAV99 replied:
@Kongkille that is why I used `document.documentElement.lang` we already set the document language with i18n in `i18n.js`.
https://github.com/electricitymap/electricitymap-contrib/blob/346d621810809ce2f7359212fb092ebeedb2a052/web/src/helpers/i18n.js#L80-L81

I have used `document.documentElement.lang` in other places but we could change those to match, either way it would be good to decide how we want to do it so the codebase is consistent.

Kongkille replied:
Well if I understand correctly, I don't think we want to rely on a document attribute in all places.

In my PR i remove this line so we no longer use this hook. This means that (because TimeAxis is a memo component) we don't actually re-render the component onLanguageChange.
https://github.com/electricitymap/electricitymap-contrib/blob/346d621810809ce2f7359212fb092ebeedb2a052/web/src/components/graph/timeaxis.jsx#L31

Essentially this means that language specific changes are not going through unless the component is guaranteed to rerender. Right now it works on master because they all somehow use the hook, but it can lead to some odd behavior.

As for the Memo I think you're right, I am not sure what the cost of it is but unless it's really expensive we should probably avoid it, if not for readability

Kongkille replied:
You can see here that the TimeAxis actually don't re-render on this PR


https://user-images.githubusercontent.com/45588799/167462682-1837d652-fa3d-4ea9-9920-4d515dec03b3.mov

So I think that we should use the "useTranslation" hook to get the language as well (or create a new hook specifically to get the languageCode perhaps)

VIKTORVAV99 replied:
As we aren't passing this global object to the function/component but rather using it directly don't this make memo unnecessary? (I haven't used react much so please correct me if I'm wrong).

Originally I figured it would be best to use a global object over function calls/imports as they would have lower performance overhead but it probably don't matter much when we only do a few calls each time. 

However if we do need memo I think it does make sense to create a languageCode hook for clarity and consistency.

Kongkille replied:
I think certainly some of our Memo components/functions are not necessary but right now we have them in the codebase, so my point is moreso that we risk the components not rerendering.

It seems to be that the most 'react' approach would be to have a languageCode hook so we absolutely guarantee that the component rerenders when language changes.

In the future we could perhaps look into some of these memos to determine how useful they are :)

madsnedergaard commented on web/src/layout/timeController.js:
```diff
@@ -0,0 +1,86 @@
+  width: calc((14vw + 16rem) - 70px); // Ensures it is smaller than countrypanel
+  z-index: 999; // TODO: find the needed one
+  position: fixed;
+  box-shadow: rgba(0, 0, 0, 0.66) 0px 10px 36px 0px, rgba(0, 0, 0, 0.06) 0px 0px 0px 1px;
```
It's a bit intense box-shadow, maybe reuse the one from `<Legend />`?

madsnedergaard commented on web/public/locales/en.json:
```diff
@@ -49,6 +49,13 @@
         "now": "Now"
     },
+    "time-controller": {
+        "title": "Display data from the last",
```
```suggestion
        "title": "Display data from the past",
```

madsnedergaard reviewed (commented):
Overall this is super cool 😍 
Some comments and ideas, but I think it works pretty well!

rasmusjeppsen commented on web/src/components/graph/timeaxis.js:
```diff
@@ -4,13 +4,29 @@ import { range } from 'lodash';
-    ? __('country-panel.now')
-    : moment(v).format('LT')
-);
+const TOTAL_TICK_COUNT = 25; // total number of ticks to be displayed
+const TICK_VALUE_FREQUENCY = 6; // Frequency at which values are displayed for a tick
```
Yes, +1! Love this change @madsnedergaard 🎉

Kongkille replied:
I've done that now!

We can also reconsider when looking into some other timeranges where there are even more ticks. We might need to filter some of them off :)

Kongkille commented on web/src/components/timeslider_new.jsx:
```diff
@@ -0,0 +1,173 @@
+import React, { useMemo, useState } from 'react';
+import { first, last, sortedIndex, isNumber } from 'lodash';
```
Excellent, just updated :)

Kongkille commented:
Thanks for the great review @VIKTORVAV99 and @madsnedergaard 🥇 

I think I've addressed your points now. As for the language items, I've just used the hook for now and then we can maybe look into more in another issue or PR?

VIKTORVAV99 commented:
It's great to see these features coming along and I'm happy to help anyway I can!

As for the discussion about hooks, yeah they aren't apart of this PR really and I think it make sense to address it at a later point.

VIKTORVAV99 commented:
looks like the build is failing because webpack has no loader configured for svg, I can add that in a separate PR if want (best option IMO) otherwise I think you need to change how you reference the svg.

madsnedergaard commented on web/src/components/timeslider_new.jsx:
```diff
@@ -0,0 +1,173 @@
+    background-position: center center;
+    background-size: 12px;
+    background-repeat: no-repeat;
+    background-image: url(${thumbImg});
```
Does it fail if we import them like we do in the stylesheets?
<img width="614" alt="Screenshot 2022-05-10 at 12 11 12" src="https://user-images.githubusercontent.com/3296643/167605600-ade3e860-6098-4e19-a07f-8466d6079deb.png">

If yes, then @VIKTORVAV99's suggestion is probably best - I guess it will in future be nice to able to use SVGs anyway :)

VIKTORVAV99 replied:
No it won't (I think), the difference here is that webpack is trying to import the contents of the SVG file (and inline it) and use that while in the stylesheet we just do "use this file" basically.

Kongkille replied:
Just in the spirit of finalizing this PR I've used the .scss file. We will have to look at this very soon again, as we'll need to remove the entire file.

Then we can handle the SVG loader in a separate PR

VIKTORVAV99 replied:
The scss files are getting removed? 👀

Last I checked the postcss don't support styled components so we would lose autoprefixer and quite a few optimizations doing that?

madsnedergaard commented on web/src/components/timeControls.js:
```diff
@@ -63,15 +64,36 @@ const DateOptionWrapper = styled.div`
-  { key: '5year', label: 'time-controller.5year' },
-];
+const getOptions = (language) =>
+  useMemo(
```
I don't think it makes sense to useMemo here - in general it should only be used for heavy calculations, as the performance overhead of using the functionality is otherwise higher than not using it :)

madsnedergaard replied:
Nevermind, you already fixed this!

madsnedergaard commented on web/src/layout/leftpanel/zonelistpanel.js:
```diff
@@ -42,10 +42,11 @@ const StyledWrapper = styled.div`
   const { __ } = useTranslation();
-  const features = useFeatureToggle();
+  const isHistoryFeatureEnabled = useFeatureToggle('history');
+  console.log(isHistoryFeatureEnabled);
```
Console.log :)

madsnedergaard commented on web/src/components/graph/timeaxis.jsx:
```diff
@@ -1,9 +1,10 @@
 import React from 'react';
+import { useTranslation } from 'react-i18next';
```
Please use our own hook instead (sorry that the naming is identical, I couldn't come up with a better name 😞)
```
import { useTranslation } from '../helpers/translation';
```
It's not critical for this use case, but it'll be annoying if we later want to use __ and it's not working as expected (with sprintf) :)

VIKTORVAV99 replied:
Maybe we should just use this?
https://react.i18next.com/latest/translation-render-prop

madsnedergaard replied:
Hmm, I personally feel like the pattern of render props is a bit out-of-date - I'd rather just either find a better name for our custom hook or add an eslint rule that ensures we only import the right one :)

VIKTORVAV99 replied:
Maybe but it would come with benefits such as inbuilt suspense for when the language file is loading and would ge guaranteed to re-render on language change.

Might be worth considering at least.

Kongkille commented:
Seems build is passing now :)

madsnedergaard commented on web/src/components/graph/timeaxis.jsx:
```diff
@@ -1,5 +1,5 @@
 import React from 'react';
-import { useTranslation } from 'react-i18next';
+import { useTranslation } from '../../helpers/translation';
```
This is also a thing in all the other places - sorry, this is the last thing before merging! 😬

madsnedergaard reviewed (approved):
🚀
